### PR TITLE
Added IsolateNameServer functionality

### PR DESCRIPTION
**Pull request by bkonyi**
Added IsolateNameServer functionality, which allows for the association of string names with isolate SendPort ids that can be used to establish inter-isolate communications.

### CODE REVIEW

File: lib/ui/isolate_name_server.dart [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+abstract class IsolateNameServer {
+  // Looks up the [SendPort] associated with a given name. Returns null
+  // if the name does not exist.
+  static SendPort lookupPortByName(String name) =>
+      _lookupPortByNameNative(name);
+
+  // Registers a SendPort with a given name. Returns true if registration is
+  // successful, false if the name entry already exists.
+  static bool registerPortWithName(SendPort port, String name) {
+    if (port == null) {
```
Comment (chinmaygarde) [inline lib/ui/isolate_name_server.dart]: Let's make sure the name valid too.

File: lib/ui/isolate_name_server.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      throw new ArgumentError("'port' cannot be null.");
+    }
+    return _registerPortWithNameNative(port, name);
```
Comment (chinmaygarde) [inline lib/ui/isolate_name_server.dart]: I don't think the convention is to add the native suffix.

File: lib/ui/isolate_name_server/isolate_name_server.cc [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+Dart_Port IsolateNameServer::LookupIsolatePortByNameUnprotected(
+    const std::string& name) {
+  auto port_iterator = port_mapping_.find(name);
+  if (port_iterator != port_mapping_.end()) {
+    return port_iterator->second;
+  }
+  return ILLEGAL_PORT;
+}
+
+bool IsolateNameServer::RegisterIsolatePortWithName(Dart_Port port,
```
Comment (chinmaygarde) [inline lib/ui/isolate_name_server/isolate_name_server.cc]: Lets use the annotations in `//flutter/fml/synchronization/thread_annotations.h` for statically enforcing threading rules.

**Comment by bkonyi** [comment]
Friendly bump @chinmaygarde

File: lib/ui/isolate_name_server/isolate_name_server.cc (hunk 3 above)
Comment (chinmaygarde) [reply lib/ui/isolate_name_server/isolate_name_server.cc]: If you define _LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS (which the default build flags already do) you get those annotations for free.

Comment (bkonyi) [reply lib/ui/isolate_name_server/isolate_name_server.cc]: Ah okay. Turns out I wasn't doing things totally right and was getting warnings as a result. Done.

File: lib/ui/isolate_name_server/isolate_name_server.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+class IsolateNameServer {
+ public:
+  IsolateNameServer() {}
+
+  // Looks up the Dart_Port associated with a given name. Returns ILLEGAL_PORT
+  // if the name does not exist.
+  Dart_Port LookupIsolatePortByName(const std::string& name)
+      LOCK_UNLOCK(mutex_);
+  ;
```
Comment (zanderso) [inline lib/ui/isolate_name_server/isolate_name_server.h]: stray ;

File: testing/dart/isolate_name_server_test.dart [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  });
+
+  test('simple isolate name server', () {
+    // Mapping for 'foobar' isn't set. Check these cases to ensure correct
+    // negative response.
+    expect(IsolateNameServer.lookupPortByName(kPortName), isNull);
+    expect(IsolateNameServer.removePortNameMapping(kPortName), isFalse);
+
+    // Register a SendPort.
+    final receivePort = new ReceivePort();
```
Comment (zanderso) [inline testing/dart/isolate_name_server_test.dart]: Do you need to close these ports?

Comment (bkonyi) [reply testing/dart/isolate_name_server_test.dart]: They're not being used for sending + receiving so I _think_ this should be fine.

**Pull request merged**